### PR TITLE
Add CSS color parsing example

### PR DESCRIPTION
**Pull request by minamorl**

### CODE REVIEW

File: examples/color.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  }
+}
+
+class Color implements HasBoundaryCheck {
+  constructor(
+    public r: RGBValue,
+    public g: RGBValue,
+    public b: RGBValue,
+    public a: Transparency | null
+  ) {}
```
Comment (wavebeem) [inline examples/color.ts:53]: I would like to focus more on the parser in this example, so maybe less class code?

Just the class Color like:

```ts
class Color {
  constructor(public r: number, public g: number, public b: number, public a: number = 1)  {}
  toString() { ... }
}
```

File: examples/color.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      `r:${this.r.value} g:${this.g.value} b:${this.b.value}` +
+      (this.a ? ` alpha:${this.a.value}` : "")
+    );
+  }
+}
+
+function repeat<T>(parser: bnb.Parser<T>, times: number) {
+  const parsers: bnb.Parser<T>[] = Array(times).fill(parser);
+  return bnb.all(...parsers);
+}
```
Comment (wavebeem) [inline examples/color.ts:71]: This is a cool parser, but I think this file would be easier to understand without it

File: examples/color.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+const sharp = bnb.text("#");
+
+const hexval = bnb.match(/[0-9a-fA-F]/);
+
+const integer = bnb.match(/[0-9.]+/);
+
+// parsers
+const pattern1 = sharp
+  .next(repeat(hexval, 3))
+  .skip(bnb.eof)
```
Comment (wavebeem) [inline examples/color.ts:83]: I don't think you need this `.skip`

File: examples/color.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  .map((values) =>
+    values.map((v) => parseInt(v + v, 16)).map((v) => new RGBValue(v))
+  )
+  .map(([r, g, b]) => new Color(r, g, b, null));
```
Comment (wavebeem) [inline examples/color.ts:87]: These two `.map` can be combined like:

```js
.map(([r, g, b]) => {
  return new Color(
    parseInt(r + r, 16),
    parseInt(g + g, 16),
    parseInt(b + b, 16),
    1
  );
});
```

File: examples/color.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+const pattern2 = sharp.next(
+  repeat(
+    repeat(hexval, 2).map((strs) => strs.join("")),
+    3
+  )
+    .map((values) =>
+      values.map((v) => parseInt(v, 16)).map((v) => new RGBValue(v))
+    )
+    .map(([r, g, b]) => new Color(r, g, b, null))
+);
```
Comment (wavebeem) [inline examples/color.ts:98]: Same idea as above refactor

File: examples/color.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+const pattern4 = bnb
+  .text("rgba(")
+  .next(integer.sepBy1(bnb.text(",")))
+  .skip(bnb.text(")"))
+  .map(([r, g, b, a]) => [...[r, g, b].map((v) => parseInt(v)), parseFloat(a)])
+  .map(([r, g, b, a]) => [
+    ...[r, g, b].map((v) => new RGBValue(v)),
+    new Transparency(a),
+  ])
+  .map(([r, g, b, a]) => new Color(r, g, b, a));
```
Comment (wavebeem) [inline examples/color.ts:117]: Easier to read with a single `.map` call I think.

File: examples/color.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+const parsers = bnb.choice(pattern1, pattern2, pattern3, pattern4);
```
Comment (wavebeem) [inline examples/color.ts:119]: Maybe new names?

```js
bnb.choice(
  hexColor1,
  hexColor2,
  rgbColor,
  rgbaColor
);
```

**Comment by minamorl** [comment]
@wavebeem 
I corrected most of codes as you reviewed. But there's still a necessity of `bnb.eof`. If I run the codes without it, the example 2 and 3 won't works. Am I missing something?

BTW:
- I like the design of bnb. I confused the difference between `parser.next` and `parser.and` but it's okay once I get used to. The documentation was very helpful.
- I want to add a `parser.repeat(n: number)` function, which I implemented in a first commit.

**Comment by wavebeem** [comment]
> But there's still a necessity of bnb.eof

I figured out the bug. hexColor2 should come before hexColor1.

```js
bnb.choice(hexColor1, hexColor2, rgbColor, rgbaColor)
```

hexColor1 and hexColor2 start out the same, so hexColor2 has to come first since it's longer.

> I like the design of bnb.
> The documentation was very helpful.

Yay! That's great :)

> I want to add a parser.repeat(n: number)

This could be a good new issue. Parsimmon has some functions like repeat that are helpful sometimes. I want to have discussions about new functions in GitHub issues so bnb doesn't get too big.

**Comment by minamorl** [comment]
I've fixed that code a bit as you pointed out.

**Comment by wavebeem** [comment]
@minamorl Does this new documentation help? https://wavebeem-bread-n-butter.netlify.app/api/#bnb.choice

**Comment by minamorl** [comment]
It helps. Thanks. You can merge this at any time you like.

**Pull request merged**